src/instructlab/sdg/eval_data.py

```diff
@@ -0,0 +1,117 @@
+# Standard
+from importlib import resources
+from typing import Any
+import re
+
+# Third Party
+from datasets import Dataset
+import yaml
+
+# First Party
+from instructlab.sdg.pipeline import EVAL_PIPELINES_PKG, Pipeline
+
+# Local
+from .logger_config import setup_logger
+
+logger = setup_logger(__name__)
+
+
+def _extract_options(text: str) -> list[Any]:
+    """regex to extract options from mcq
+
+    Args:
+        text (str): question with options/mcq choices
+
+    Returns:
+        list[Any]: options under question that match the pattern.
+    """
+    # Use a regular expression to find patterns and capture the text after the letter and parenthesis
+    pattern = r"\b[A-Z]\) (.+)"
+    matches = re.findall(pattern, text)
+    return matches
+
+
+def _format_mmlu_style(ds: Dataset) -> Dataset:
+    """Format the dataset according to lm-harness mmlu requirement.
+
+    Args:
+        ds (Dataset): input dataset
+
+    Returns:
+        Dataset: formated hf dataset
+    """
+    ds = ds.map(
+        lambda x: {"answer": x["mmlubench_answer"][: x["mmlubench_answer"].index(")")]}
+    )
+    ds = ds.map(lambda x: {"choices": _extract_options(x["mmlubench_question"])})
+    ds = ds.map(
+        lambda x: {
+            "question": x["mmlubench_question"][
+                : x["mmlubench_question"].index("A)")
+            ].strip()
+        }
+    )
+    ds = ds.rename_columns({"domain": "subject"})
+    ds = ds.filter(lambda x: x["choices"])
+    ds = ds.filter(lambda x: len(x["choices"]) == 4)
+    ds = ds.filter(lambda x: x["answer"] in ["A", "B", "C", "D"])
+    ds = ds.class_encode_column("answer")
+    return ds
+
+
+def _post_process_mcq(ds: Dataset) -> Dataset:
+    """Filter, format and return Multiple Choice Question (MCQ) dataset that
+    is compatible with lm-harness for doing mmlu-style evaluation
+
+    Filters out badly generated data, adds dataset type column
+
+    Args:
+        ds (Dataset): mcq generated dataset from mmmlu pipeline
+
+    Returns:
+        Dataset: Hf Dataset with new column, filtered dataset
+    """
+    ds = ds.filter(lambda x: ")" in x["mmlubench_answer"])
+    ds = ds.filter(lambda x: "A)" in x["mmlubench_question"])
+    ds = ds.add_column("dataset_type", ["mcq_qa"] * ds.num_rows)
+    return _format_mmlu_style(ds)
+
+
+def _create_mmlu_evaluation_task(task_name, eval_data_file_path, yaml_file_path):
+    """
+    Prepare Task Yaml that will be used in lm_eval_harness to evaluate knowledge using mmlu style metric
+    """
+    task_yaml = {
+        "task": task_name,
+        "dataset_kwargs": {"data_files": {"test": eval_data_file_path}},
+        "include": "_default_mmlu_pr_template_yaml",
```

Comment:
I'm not entirely clear where this `_default_mmlu_pr_template_yaml` lives - the eval library?

I'm definitely nervous about tight coupling between the libraries - this sort of thing can make it difficult to make changes

Might it make sense to put `_default_mmlu_pr_template_yaml` in the same codebase that generates this task?

Reply:
is this the file its referencing ? https://github.com/instructlab/instructlab/blob/main/tests/testdata/mmlu_branch/_default_mmlu_pr_template_yaml

Reply:
Oh, sorry - I thought I linked to that. Yes, that's what it looks like

Thinking about it afterwards ... and looking at https://github.com/EleutherAI/lm-evaluation-harness/blob/main/docs/task_guide.md#including-a-base-yaml

Why not just include all that task config in the file we write out? There's no huge benefit to referencing another file and all the coordination that comes with that?

Reply:
Makes sense looks like part of it is included here already, may aswell have it all